src/Lib.P2P/SwarmService.cs

```diff
@@ -0,0 +1,1132 @@
+            }
+
+            peer = RegisterPeer(peer);
+
+            // If connected and still open, then use the existing connection.
+            if (Manager.TryGet(peer, out var conn))
+            {
+                return conn;
+            }
+
+            // Use a current connection attempt to the peer or create a new one.
+            try
+            {
+                using (var cts = CancellationTokenSource.CreateLinkedTokenSource(_swarmCancellation.Token, cancel))
+                {
+                    return await _pendingConnections
+                       .GetOrAdd(peer,
+                            key => new AsyncLazy<PeerConnection>(() => DialAsync(peer, peer.Addresses, cts.Token)))
+                       .ConfigureAwait(false);
+                }
+            }
+            catch (Exception)
+            {
+                PeerNotReachable?.Invoke(this, peer);
+                throw;
+            }
+            finally
+            {
+                _pendingConnections.TryRemove(peer, out _);
+            }
+        }
+
+        /// <summary>
+        ///   Create a stream to the peer that talks the specified protocol.
+        /// </summary>
+        /// <param name="peer">
+        ///   The remote peer.
+        /// </param>
+        /// <param name="protocol">
+        ///   The protocol name, such as "/foo/0.42.0".
+        /// </param>
+        /// <param name="cancel">
+        ///   Is used to stop the task.  When cancelled, the <see cref="TaskCanceledException"/> is raised.
+        /// </param>
+        /// <returns>
+        ///   A task that represents the asynchronous operation. The task's result
+        ///   is the new <see cref="Stream"/> to the <paramref name="peer"/>.
+        /// </returns>
+        /// <remarks>
+        ///   <para>
+        ///   When finished, the caller must <see cref="Stream.Dispose()"/> the
+        ///   new stream.
+        ///   </para>
+        /// </remarks>
+        public async Task<Stream> DialAsync(Peer peer,
+            string protocol,
+            CancellationToken cancel = default)
+        {
+            peer = RegisterPeer(peer);
+
+            // Get a connection and then a muxer to the peer.
+            var connection = await ConnectAsync(peer, cancel).ConfigureAwait(false);
+            var muxer = await connection.MuxerEstablished.Task.ConfigureAwait(false);
+
+            // Create a new stream for the peer protocol.
+            var stream = await muxer.CreateStreamAsync(protocol, cancel).ConfigureAwait(false);
+            try
+            {
+                await connection.EstablishProtocolAsync("/multistream/", stream, cancel).ConfigureAwait(false);
+
+                await Message.WriteAsync(protocol, stream, cancel).ConfigureAwait(false);
+                var result = await Message.ReadStringAsync(stream, cancel).ConfigureAwait(false);
+                if (result != protocol) throw new Exception($"Protocol '{protocol}' not supported by '{peer}'.");
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051852)